sandbox/apparmor/notify/message.go

```diff
@@ -0,0 +1,418 @@
+	if !msg.ModeSet.IsValid() {
+		return fmt.Errorf("unsupported modeset: %d", msg.ModeSet)
+	}
+	return nil
+}
+
+// MsgNotification describes a kernel notification message.
+//
+// This structure corresponds to the kernel type struct apparmor_notif
+// described below.
+//
+//	struct apparmor_notif {
+//	  struct apparmor_notif_common base;
+//	  __u16 ntype;        /* notify type */
+//	  __u8 signalled;
+//	  __u8 flags;
+//	  __u64 id;           /* unique id, not globally unique*/
+//	  __s32 error;        /* error if unchanged */
+//	} __attribute__((packed));
+type MsgNotification struct {
+	MsgHeader
+	// NotificationType describes the kind of notification message used.
+	// Currently the kernel only sends APPARMOR_NOTIF_OP messages.
+	NotificationType NotificationType
+	// XXX: Signaled seems to be unused.
+	Signalled uint8
+	// Set to 1 to NOT cache
+	Flags uint8
+	// ID is an opaque kernel identifier of the notification message. It must be
+	// repeated in the MsgNotificationResponse if one is sent back.
+	ID uint64
+	// XXX: This seems to be unused and clashes with identical field in
+	// MsgNotificationOp.
+	Error int32
+}
+
+// UnmarshalBinary unmarshals the message from binary form.
+func (msg *MsgNotification) UnmarshalBinary(data []byte) error {
+	const prefix = "cannot unmarshal apparmor notification message"
+
+	// Unpack the base structure.
+	if err := msg.MsgHeader.UnmarshalBinary(data); err != nil {
+		return err
+	}
+
+	// Unpack fixed-size elements.
+	buf := bytes.NewBuffer(data)
+	order := arch.Endian() // ioctl messages are native byte order, verify endianness if using for other messages
+	if err := binary.Read(buf, order, msg); err != nil {
+		return fmt.Errorf("%s: cannot unpack: %s", prefix, err)
+	}
+
+	return nil
+}
+
+// MarshalBinary marshals the message into binary form.
+func (msg *MsgNotification) MarshalBinary() ([]byte, error) {
+	msg.Version = 2
+	msg.Length = uint16(binary.Size(*msg))
+	buf := bytes.NewBuffer(make([]byte, 0, msg.Length))
+	order := arch.Endian() // ioctl messages are native byte order, verify endianness if using for other messages
+	if err := binary.Write(buf, order, msg); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+// Validate returns an error if the mssage contains invalid data.
+func (msg *MsgNotification) Validate() error {
+	if !msg.NotificationType.IsValid() {
+		return fmt.Errorf("unsupported notification type: %d", msg.NotificationType)
+	}
+	return nil
+}
+
+// MsgNotificationResponse describes a response to a MsgNotification.
+//
+// This structure corresponds to the kernel type struct apparmor_notif
+// described below.
+//
+//	struct apparmor_notif_resp {
+//	  struct apparmor_notif base;
+//	  __s32 error;        /* error if unchanged */
+//	  __u32 allow;
+//	  __u32 deny;
+//	} __attribute__((packed));
+type MsgNotificationResponse struct {
+	MsgNotification
+	// XXX: The embedded MsgNotification also has an Error field, why?
```

Comment:
Is this still the case in the underlying apparmor structs as well?

Reply:
The ioctl interface use some common structs shared/embedded in other structs. In this case we have
```
struct __attribute__((__packed__)) apparmor_notif_common {
	__u16 len;			/* actual len data */
	__u16 version;			/* interface version */
};

struct __attribute__((__packed__)) apparmor_notif {
	struct apparmor_notif_common base;
	__u16 ntype;			/* notify type */
	__u8 signalled;
	__u8 reserved;
	__u64 id;			/* unique id, not gloablly unique*/
	__s32 error;			/* error if unchanged */
};
```

the error field contains the error the kernel will return to the application if the notification is denied.